File: csharp/ReceiptPrinter.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;

namespace supermarket
{
    public class ReceiptPrinter {

    private int columns;
    private static CultureInfo  culture = CultureInfo.CreateSpecificCulture("en-GB");


    public ReceiptPrinter(int columns) {
        this.columns = columns;
    }

    public string printReceipt(Receipt receipt) {
        StringBuilder result = new StringBuilder();
        foreach (ReceiptItem item in receipt.getItems())
        {
            string price = item.getTotalPrice().ToString("F", culture);
            string quantity = presentQuantity(item);
            string name = item.getProduct().getName();
            string unitPrice = item.getPrice().ToString("F", culture);

            int whitespaceSize = this.columns - name.Length - price.Length;
            string line = name + getWhitespace(whitespaceSize) + price + "\n";

            if (item.getQuantity() != 1) {
                line += "  " + unitPrice + " * " + quantity + "\n";
            }
            result.Append(line);
        }
        foreach (Discount discount in receipt.getDiscounts()) {
            string productPresentation = discount.getProduct().getName();
            string pricePresentation = discount.getDiscountAmount().ToString("F", culture);
            string description = discount.getDescription();
            result.Append(description);
            result.Append("(");
            result.Append(productPresentation);
            result.Append(")");
            result.Append(getWhitespace(this.columns - 3 - productPresentation.Length - description.Length - pricePresentation.Length));
            result.Append("-");
            result.Append(pricePresentation);
            result.Append("\n");
        }

        {
            result.Append("\n");
            string pricePresentation = receipt.getTotalPrice().ToString("F", culture);
            string total = "Total: ";
            string whitespace = getWhitespace(this.columns - total.Length - pricePresentation.Length);
            result.Append(total).Append(whitespace).Append(pricePresentation);

            }
            return result.ToString();
    }

    private static string presentQuantity(ReceiptItem item) {
        return ProductUnit.Each == item.getProduct().getUnit()
                ? ((int)item.getQuantity()).ToString()
                : item.getQuantity().ToString("N3", culture);
    }

    private static string getWhitespace(int whitespaceSize) {
        StringBuilder whitespace = new StringBuilder();
        for (int i = 0; i < whitespaceSize; i++) {
            whitespace.Append(" ");
        }
        return whitespace.ToString();
    }
}

}

File: php/src/ReceiptPrinter.php
<?php

declare(strict_types=1);

namespace Supermarket;

use Supermarket\Model\Discount;
use Supermarket\Model\ProductUnit;
use Supermarket\Model\Receipt;
use Supermarket\Model\ReceiptItem;

class ReceiptPrinter
{
    public function __construct(
        private int $columns = 40
    ) {
    }

    public function printReceipt(Receipt $receipt): string
    {
        $result = '';
        foreach ($receipt->getItems() as $item) {
            $itemPresentation = $this->presentReceiptItem($item);
            $result .= $itemPresentation;
        }

        foreach ($receipt->getDiscounts() as $discount) {
            $discountPresentation = $this->presentDiscount($discount);
            $result .= $discountPresentation;
        }

        $result .= "\n";
        $result .= $this->presentTotal($receipt);
        return $result;
    }

    protected function presentReceiptItem(ReceiptItem $item): string
    {
        $price = self::presentPrice($item->getTotalPrice());
        $name = $item->getProduct()->getName();

        $line = $this->formatLineWithWhitespace($name, $price) . "\n";

        if ($item->getQuantity() !== 1.0) {
            $line .= '  ' . self::presentPrice($item->getPrice()) . ' * ' . self::presentQuantity($item) . "\n";
        }
        return $line;
    }

    protected function presentDiscount(Discount $discount): string
    {
        $name = "{$discount->getDescription()}({$discount->getProduct()->getName()})";
        $value = self::presentPrice($discount->getDiscountAmount());

        return $this->formatLineWithWhitespace($name, $value) . "\n";
    }

    protected function presentTotal(Receipt $receipt): string
    {
        $name = 'Total: ';
        $value = self::presentPrice($receipt->getTotalPrice());
        return $this->formatLineWithWhitespace($name, $value);
    }

    protected function formatLineWithWhitespace(string $name, string $value): string
    {
        $whitespaceSize = $this->columns - strlen($name) - strlen($value);
        return $name . str_repeat(' ', $whitespaceSize) . $value;
    }

    protected static function presentPrice(float $price): string
    {
        return sprintf('%.2F', $price);
    }

    private static function presentQuantity(ReceiptItem $item): string
    {
        return $item->getProduct()->getUnit()->equals(ProductUnit::EACH()) ?
            sprintf('%x', $item->getQuantity()) :
            sprintf('%.3F', $item->getQuantity());
    }
}
